File: php/jobs.php
<?php
require_once 'entity.php';
require_once 'users.php';
require_once 'products.php';
require_once 'job_comments.php';
require_once 'draft_comments.php';
require_once 'production_comments.php';
require_once 'country_taxes.php';
require_once 'domains.php';
require_once 'drafts.php';
require_once 'invoices.php';
require_once 'files.php';
require_once 'addresses.php';
require_once 'bids.php';
require_once 'shipments.php';
require_once 'companies.php';
require_once 'specifications.php';
require_once 'notifications.php';

require_once './../php_aux/business_default.php';
require_once './../php_aux/roles.php';
require_once './../php_aux/status.php';
require_once './../php_aux/notification_sources.php';
require_once './../php_aux/events.php';

class Job extends Entity
{
    public static $json_name = 'Job';
    public static $resource = '/jobs/';


    public function __construct() {
        $this->escape_fields = ['product', 'quantity', 'tax_type'];
        $this->json_property('id', 'integer');
        $this->json_property('manager','User', $many = False,
                             $recursive = True);
        $this->json_property('designer','User', $many = False,
                             $recursive = True);
        $this->json_property('client','User', $many = False,
                             $recursive = True);
        $this->json_property('product','Product', $many = False,
                             $recursive = True);
        $this->json_property('comments','JobComment', $many = True,
                             $recursive = True);
        $this->json_property('draft_comments','DraftComment', $many = True,
                             $recursive = True);
        $this->json_property('drafts','Draft', $many = True,
                             $recursive = True);
        $this->json_property('invoice','Invoice', $many = False,
                             $recursive = True);
        $this->json_property('shipping','Address', $many = False,
                             $recursive = True);
        $this->json_property('production_shipping_address','Address', $many = False,
                             $recursive = True);
        $this->json_property('notifications','Notification', $many = True,
                             $recursive = True);
        $this->json_property('client_company','Company', $many = False,
                             $recursive = True);
        $this->json_property('quantity', 'integer');
        $this->json_property('notes', 'string');
        $this->json_property('production_notes', 'string');
        $this->json_property('completed', 'boolean');
        $this->json_property('archived', 'boolean');
        $this->json_property('priority', 'integer');
        $this->json_property('cost_per_unit', 'float');
        $this->json_property('tax_amount', 'float');
        $this->json_property('automatic_price_enabled', 'boolean');
        $this->json_property('cost', 'integer');
        $this->json_property('job_weight', 'float');
        $this->json_property('job_volume', 'float');
        $this->json_property('needs_production', 'boolean');
        $this->json_property('needs_drafting', 'boolean');
        $this->json_property('needs_shipping', 'boolean');
        $this->json_property('domain', 'string');
        $this->json_property('production_status', 'integer');
        $this->json_property('design_status', 'integer');
        $this->json_property('shipping_status', 'integer');
        $this->json_property('payment_status', 'integer');
        $this->json_property('domain', 'string');
        $this->json_property('received', 'DateTime');
        $this->json_property('deadline', 'DateTime');
        $this->json_property('updated', 'DateTime');
        $this->json_property('product', 'Product', $many = False,
                             $default = '1', $recursive = True);
        $this->json_property('domain', 'Domain', $many = False,
                             $default = '1', $recursive = True);
        $this->json_property('clientFile', 'File', $default = [],
                             $many = true, $recursive = True);
        $this->json_property('productionFile', 'File', $default = [],
                             $many = true, $recursive = True);
        $this->json_property('assignments', 'Assignment', $default = [],
                             $many = true, $recursive = True);
        $this->json_property('specifications_groups', 'SpecificationsGroup', $default = [],
                             $many = true, $recursive = True);
        $this->json_property('specifications', 'Specification', $default = [],
                             $many = true, $recursive = True);
        $this->json_property('shipment', 'Shipment',
                             $many = False, $recursive = True);
        $this->json_property('tax_type', 'CountryTax',
                             $many = False, $recursive = True);
    }

    function product_total_cost(){
        /*Return the cost_per_unit multiplied by the total job
            quantity
        */
        return $this->quantity * $this->cost_per_unit;
    }

    function process_for_transfer(){
        $this->updated = NULL;
    }

    function is_draft_accepted(){
        /*Return True if the current draft for this job has been accepted,
            or False otherwise.
        */
        try {
            $current_draft = $this->drafts[0];
        } catch (Exception $e) {
            return False;
        }
        if ($current_draft->accepted){
            return True;
        }else {
            return False;
        }
    }

    function priority_name(){
        /*Return a string from one of the constants defined in
            common.business_default that represents the priority of this job.
        */
        global $PRIORITY_OPTIONS_REVERSE_MAP;
        $return_value = array_key_exists($this->priority, $PRIORITY_OPTIONS_REVERSE_MAP)
        ? $PRIORITY_OPTIONS_REVERSE_MAP[$this->priority] : LOW_STRING;
        return $return_value;
    }

    function current_draft(){
        #Return the most recent job draft from an array of job drafts
        if(sizeof($this->drafts) > 0){
            return $this->drafts[0];
        } else {
            return NULL;
        }
    }

    function most_recent_draft_id(){
        /*Return the current draft id for the job or None if no drafts.
            Requires that the draft objects be embedded when fetching
            this job object.
        */
        if (sizeof($this->drafts == 0)){
            return NULL;
        } else{
            return $this->drafts[0]->id;
        }
    }

    function supplier($only_aproved = False){
        /*Return either a list of suppliers, a single supplier, or None,
            depending on what stage the job is in and what suppliers have
            been assigned. Requires that assignments.supplier be embedded
            with the job when fetching it.
        */
        if($this->assignments){
            if($only_aproved){
                foreach($this->assignments as $assign){
                    if($assign->manager_accepts){
                        return $assign->supplier;
                    }
                }
            }else {
                $return_val = array();
                foreach($this->assignments as $assign){
                    $return_val[] = $assign->supplier;
                }
                return $return_val;
            }
        }
    }

    function is_production_in_house(){
        #Return True if one of the job assignment suppliers is a manager
        foreach($this->assignments as $assign){
            if($assign->supplier->has_authority($this->domain->id,[ADMIN, MANAGER])){
                return True;
            }
        }
        return False;
    }

    function in_house_assignment(){
        #Return an assignment if the assignment is in-house
        if($this->is_production_in_house()){
            return $this->assignments[0];
        }
        return NULL;
    }

    function accepted_assignment(){
        #Return the production assignment accepted by the manager
        foreach($this->assignments as $assign){
            if($assign->manager_accepts){
                return $assign;
            }
        }
        return NULL;
    }

    function assignment_deadline(){
        #Return the assignment deadline from one of the assignments
        if(sizeof($this->assignments) > 0){
            return $this->assignments[0]->assignment_deadline;
        }
        return NULL;
    }

    function assignment_by_supplier_id($user_id){
        /*Return a job assignment associated with
            the user provided and return None if no assignment can
            be found with the user.
        */
        $assignment  = NULL;
        foreach($this->assignments as $assign){
            if($assign->supplier->id === $user_id){
                $assignment = $assign;
            }
        }
        return $assignment;
    }

    function production_shipment(){
        /*Return the shipment object which has been created for production
            procedure.
        */
        foreach($this->assignments as $assign){
            if($assign->shipment){
                return $assign->shipment;
            }
        }
        return NULL;
    }

    function production_display_info(){
        /*Return the PRODUCTION_STATUS dict from
            common.status with the values based on the job.production_status
        */
        return status_info($this->design_status, DESIGN_STATUS, "design ");
    }

    function payment_display_info(){
        /*Return the PAYMENT_STATUS dict from
            common.status with the values based on the job.payment_status
        */
        return status_info($this->payment_status, PAYMENT_STATUS, "payment ");
    }

    function shipment_display_info(){
        /*Return the SHIPPING_STATUS dict from common.status with the values
            based on the job.shipping_status
        */
        return status_info($this->shipping_status, SHIPPING_STATUS, "shipment ");
    }

    function supplier_shipment_info(){
        #Return the shipping information submitted buy the supplier
        $assignment = $this->accepted_assignment();
        if($assignment){
            return $assignment->shipment;
        }
        return NULL;
    }

    function product_unit_price(){
        /*Display the product unit_price. This value takes into consideration
            discounted unit price based on server calculated product unit price
            after it got discounted
        */
        return $this->cost_per_unit;
    }

    function assignment_earliest_production(){
        /* Return the assignment which is able to get the
          production completed in the fastest amount of time
        */
        $earliest_deadline = NULL;
        $earliest_supplier = NULL;
        foreach($this->assignments as $assign){
            if($assign->bid){
                $agreed_deadline = $assign->bid->agreed_deadline;
                if(!$earliest_deadline or $agreed_deadline < $earliest_deadline){
                    $earliest_deadline = $assign->bid->agreed_deadline;
                    $earliest_supplier = $assign;
                }
            }
        }
        return $earliest_supplier;
    }

    function assignment_lowest_bid(){
        #Return the lowest bid out of all the assignments
        $lowest_bid = NULL;
        $lowest_supplier = NULL;
        foreach($this->assignments as $assign){
            if($assign->bid){
                if(!$lowest_bid or $assign->bid->bid_total() < $lowest_bid){
                    $lowest_bid = $assign->bid->bid_total();
                    $lowest_supplier = $assign;
                }
            }
        }
        return $lowest_supplier;
    }

    function get_section_name($section){
        /*Generate section name that can be used to generate private attribute
            to store notifications info of different sections in job.
        */
        $fmt = "_%s_notifications_";
        if(!$section){
            return sprintf($fmt, "all");
        }
        return sprintf($fmt, JOB_SECTION_STRINGS[$section]);
    }

    function inject_notifications($tab="unread", $section = NULL){
        $notifications_embed = ['domain' => ['logo' => []],
                                'relatedJob' => [],
                                'sender' => ['profilePicture' => []]];
        $query = ["tab" => $tab,
                  "related_job" => $this->id,
                  "section" => $section,
                  "sort" => "id",
                  "order" => "desc"];
        global $notifications;
        list($notifications_list, $notifications_statistics) =
            $notifications->fetch($query, $embed=$notifications_embed);
        $section_prefix = $this->get_section_name($section);

        $attr_tab = $section_prefix . "tab";
        $this->$attr_tab = $tab;

        $attr_statistics = $section_prefix . "statistics";
        $this->$attr_statistics = $notifications_statistics;

        $attr_list = $section_prefix . "list";
        $this->$attr_list = $notifications_list;
    }

    function need_to_update_notifications($tab, $section = null){
        /*Check whether we need to update the job related notifications,
            if needed return True otherwise return False.
        */
        $section_prefix = get_section_name($section);

        $section_tab = $section_prefix."tab";
        $tab_attr = property_exists($this, $section_tab)?
        $this->$section_tab:null;

        $section_statistics = $section_prefix."statistics";
        $statistics_attr = property_exists($this, $section_statistics)?
        $this->$section_statistics:null;

        $section_list = $section_prefix."list";
        $list_attr = property_exists($this, $section_list)?
        $this->$section_list:null;

        return ($tab_attr != $tab) or ($statistics_attr === null) or
        ($list_attr === null);
    }

    function refresh_notifications_info($tab, $section = null){
        #Refresh the notifications of this job if needed
        if($this->need_to_update_notifications($tab, $section)){
            $this->inject_notifications($tab, $section);
        }
    }

    function notifications_statistics($tab = "unread", $section = null){
        /*Return the statistics of notifications of the this job
            whose recipient is current user.
        */
        $this->refresh_notifications_info($tab, $section);
        $section_statistics = $this->get_section_name($section)."statistics";
        try {
            return $this->$section_statistics;
        } catch (Exception $e) {
            return $e;
        }
    }

    function notifications_list($tab = "unread"){
        /*Return the count of notifications of the this job
            whose recipient is current user.
        */
        $this->refresh_notifications_info($tab);
        return $this->_all_notifications_list;
    }

    function unseen_notifications_count($section = null){
        /*Return the count of notifications of the this job
            whose recipient is current user.
        */
        $statistics =  $this->notifications_statistics($tab ="unread", $section);
        if($statistics instanceof Exception) {
            return $statistics;
        }
        return $statistics->available;
    }

    function shipping_finished(){
        #Return whether the shipping has been finished
        return $this->shiping_status >= SHIPPING_STATUS["DISPATCHED"]["dbValue"]
         or (!$this->needs_shipping);
    }

    function drafting_finished(){
        #Return whether the drafting has been finished
        return $this->design_status >= DESIGN_STATUS["DRAFTING_APPROVED"]["dbValue"]
         or (!$this->needs_drafting);
    }

    function production_finished(){
        #Return whether the production has been finished
        return isset($this->production_status) and
        $this->production_status >= PRODUCTION_STATUS["SHIPPED"]["dbValue"]
         or (!$this->needs_production);
    }

    function payment_finished(){
        #Return whether the payment has been finished
        if($this->payment_status){
            return $this->payment_status >=
            PAYMENT_STATUS["FULLY_PAID"]["dbValue"];
        }
    }

    function is_ready_to_complete(){
        /*Return whether the job is ready for move to the
            completed stage
        */
        return $this->drafting_finished() and $this->production_finished() and
         $this->payment_finished() and $this->shipping_finished();
    }

    function array_of_related_address_names_and_ids(
        $job_shipping = True, $production_shipping = true,
        $manager_addresses = true, $client_addresses = true){
        #Return a list of the related job adress ids and names.
        $saved_addresses = array();
        if($this->shipping and $job_shipping){
            $saved_addresses[] =
            ["name" => "Job Shipping Address",
            "id " => $this->production_shipping_address->id];
        }

        if($this->client and $client_addresses){
            $saved_addresses = $saved_addresses +
            $this->client->array_of_addresses_names_and_ids("Client Address");
        }

        if($this->manager and $manager_addresses){
            $saved_addresses = $saved_addresses +
            $this->manager->array_of_addresses_names_and_ids("Manager Address");
        }

        return $saved_addresses;
    }

    function draft_time_line($reverse = true){
        /*
            Return job draft related object in chronological order
            as a dictionary. This is used for creating a job draft timeline.
        */
        $time_line_array = array();

        foreach($this->draft_comments as $comment){
            $time_line_array[$comment->date] = new Event(DRAFT_COMMENT, $comment);
        }
        unset($comment);

        foreach($this->drafts as $draft){
            $time_line_array[$draft->date] = new Event(DRAFT_UPLOADED, $draft);

            if($draft->accepted){
                $time_line_array[$draft->accepted] = new Event(DRAFT_APPROVED, $draft->accepted);
            }

            if($draft->resend_date){
                $time_line_array[$draft->resend_date] = new Event(DRAFT_RESENT, $draft->resend_date);
            }

            foreach($draft->comments as $comment){
                if($comment->change_request){
                    $time_line_array[$comment->date] = new Event(DRAFT_REJECTED, $comment);
                } else {
                    $time_line_array[$comment->date] = new Event(DRAFT_COMMENT, $comment);
                }
            }
            unset($comment);
        }
        unset($draft);

        function cmp($a, $b){
            if($a == $b) {
                return 0;
            }
            return $a < $b ? -1 : 1;
        }

        function cmp_reverse($a, $b){
            if($a == $b) {
                return 0;
            }
            return $a < $b ? 1 : -1;
        }

        if($reverse) {
            uksort($time_line_array, "cmp_reverse");
            return $time_line_array;
        } else {
            uksort($time_line_array, "cmp");
            return $time_line_array;
        }
    }

    function job_info_time_line($reverse = true) {
        /*
         * Return the job comments in reverse with the youngest at
            bottom of the list.
         */
        $time_line_array = [];
        foreach($this->comments as $comment){
            $time_line_array[$comment->date] = new Event(JOB_COMMENT, $comment);
        }
        unset($comment);

        function cmp($a, $b){
            if($a == $b) {
                return 0;
            }
            return $a < $b ? -1 : 1;
        }

        function cmp_reverse($a, $b){
            if($a == $b) {
                return 0;
            }
            return $a < $b ? 1 : -1;
        }

        if($reverse) {
            uksort($time_line_array, "cmp_reverse");
            return $time_line_array;
        } else {
            uksort($time_line_array, "cmp");
            return $time_line_array;
        }
    }

    function is_complete_string(){
        /*
            Return a string 'Completed' if the job is complete
            else it will return a string 'Uncompleted'.
        */
        if($this->completed){
            return "Completed";
        }
        return "Uncompleted";
    }

    function can_be_archived_by(){
        /*Check if the job can be archived by the user making the request.*/
        return $this->rights->delete;
    }

    function total_inc_tax(){
        /*Return the job cost + the tax amount*/
        return $this->tax_amount + $this->cost;
    }
}

class Jobs extends Resource
{
    public function __construct()
    {
        $this->entity_class = 'Job';
        $this->json_name = 'jobs';
    }
}

class Assignment extends Entity
{
    public static $resource = '/assignments/';
    public static $json_name = 'assignment';

    public function __construct()
    {
        $this->json_property('id', 'integer');
        $this->json_property('manager_accepts', 'string');
        $this->json_property('supplier_refused', 'string');
        $this->json_property('production_deadline', 'DateTime');
        $this->json_property('assignment_deadline', 'DateTime');
        $this->json_property('job', 'Job',
                             $many = False, $recursive = True);
        $this->json_property('supplier', 'User',
                             $many = False, $recursive = True);
        $this->json_property('bid', 'Bid',
                             $many = False, $recursive = True);
        $this->json_property('comments', 'ProductionComment',
                             $many = True, $recursive = True);
        $this->json_property('shipment', 'Shipment',
                             $many = False, $recursive = True);
        $this->json_property('notifications', 'Notification',
                             $many = True, $recursive = True);
    }
}

class Assignments extends Resource
{
    public function __construct()
    {
        $this->entity_class = 'Assignment';
        $this->json_name = 'assignments';
    }
}

$jobs = new Jobs();
$assignments = new Assignments();

File: python/jobs.py
# pylint: disable=E0203
import collections
import datetime
import sdk.python.entities
from sdk.python.job_comments import JobComment
from sdk.python.draft_comments import DraftComment
from sdk.python.country_taxes import CountryTax
from sdk.python.domains import Domain
from sdk.python.domain_tags import DomainTag
from sdk.python.drafts import Draft
from sdk.python.invoices import Invoice
from sdk.python.files import File
from sdk.python.addresses import Address
from sdk.python.shipments import Shipment
from sdk.python.companies import Company
from sdk.python.phone_numbers import PhoneNumber
from sdk.python.email_addresses import EmailAddress
from sdk.python.variations import Variation, VariationsGroup
from sdk.python.notifications import Notification
from sdk.python.util.status import PRODUCTION_STATUS, PAYMENT_STATUS, \
    DESIGN_STATUS, SHIPPING_STATUS, status_info
from sdk.python.util import events
from sdk.python.util.roles import MANAGER, ADMIN
from sdk.python.util.notification_sources import JOB_SECTION_STRINGS
from sdk.python.util.business_default import PRIORITY_OPTIONS_REVERSE_MAP, \
    LOW_STRING
from sdk.python.entities import Property


class Job(sdk.python.entities.Entity):

    resource = '/jobs/'
    json_name = 'job'

    def __init__(self):
        super(Job, self).__init__()
        self.escape_fields = ['product', 'quantity', 'tax_type']

    id = Property(int)
    currency = Property(str)
    client = Property('sdk.python.users.User')
    manager = Property('sdk.python.users.User')
    designer = Property('sdk.python.users.User')
    product = Property('sdk.python.products.Product')
    comments = Property(JobComment)
    pre_draft_comments_count = Property(int)
    draft_comments = Property(DraftComment)
    drafts = Property(Draft)
    invoice = Property(Invoice, backref="jobs")
    domain = Property(Domain)
    shipping = Property(Address)
    production_shipping_address = Property(Address)
    notifications = Property(Notification)
    client_phone = Property(PhoneNumber)
    client_email = Property(EmailAddress)
    client_company = Property(Company)
    client_company_phone = Property(PhoneNumber)
    client_company_email = Property(EmailAddress)
    quantity = Property(int)
    notes = Property(str)
    purpose = Property(int)
    call_to_actions = Property(str)
    call_to_action_details = Property(list)
    production_notes = Property(str)
    group_buy_status = Property(int)
    production_status = Property(int)
    design_status = Property(int)
    payment_status = Property(int)
    shipping_status = Property(int)
    completed = Property(bool)
    job_info_approved_by_client = Property(bool)
    priority = Property(int)
    production_files = Property(File)
    client_files = Property(File)
    archived = Property(datetime.datetime)
    received = Property(datetime.datetime)
    deadline = Property(datetime.datetime)
    updated = Property(datetime.datetime)
    inventory_status = Property(int)
    cost_per_unit = Property(float)
    automatic_price_enabled = Property(bool)
    tax_amount = Property(float)
    cost = Property(float)
    total_cost = Property(float)
    job_weight = Property(float)
    job_volume = Property(float)
    needs_group_buy = Property(bool)
    needs_production = Property(bool)
    needs_invoicing = Property(bool)
    needs_drafting = Property(bool)
    needs_shipping = Property(bool)
    needs_inventory = Property(bool)
    deduction_date = Property(datetime.datetime)
    variations_groups = Property(VariationsGroup)
    variations = Property(Variation)
    quote_set = Property(bool)
    drop_ship = Property(bool)
    shipment = Property(Shipment, backref="jobs")
    tax_type = Property(CountryTax)
    tags = Property(DomainTag, backref="jobs")
    comments = Property(JobComment, backref="job")
    drafts = Property(Draft, backref="job")
    draft_comments = Property(DraftComment, backref="job")

    # not embedded by default, must be requested
    in_stock = Property(bool)
    limited_stock = Property(bool)
    unread_notifications_count = Property(int)
    unread_job_info_notifications_count = Property(int)
    unread_job_drafting_notifications_count = Property(int)
    unread_job_production_notifications_count = Property(int)
    unread_job_shipping_notifications_count = Property(int)
    unread_job_invoicing_notifications_count = Property(int)

    notifications = Property(Notification, backref="related_job")

    def product_total_cost(self):
        """ Return the cost_per_unit multiplied by the total job
            quantity
        """
        return self.quantity * self.cost_per_unit

    def process_for_transfer(self):
        self.updated = None  # type: ignore

    def is_draft_accepted(self):
        """ Return True if the current draft for this job has been accepted,
            or False otherwise.
        """
        try:
            current_draft = self.drafts[0]
        except IndexError:
            return False
        if current_draft.accepted:
            return True
        return False

    def priority_name(self):
        """ Return a string from one of the constants defined in
            sdk.python.util.business_default that represents the priority of
            this job.
        """
        return PRIORITY_OPTIONS_REVERSE_MAP.get(self.priority, LOW_STRING)

    def current_draft(self):
        """ Return the most recent job draft from an array of job drafts """
        if len(self.drafts) > 0:
            return self.drafts[0]
        return None

    def most_recent_draft_id(self):
        """ Return the current draft id for the job or None if no drafts.
            Requires that the draft objects be embedded when fetching
            this job object.
        """
        if len(self.drafts) == 0:
            return None
        return self.drafts[0].id

    def supplier(self, only_approved=False):
        """ Return either a list of suppliers, a single supplier, or None,
            depending on what stage the job is in and what suppliers have
            been assigned. Requires that assignments.supplier be embedded
            with the job when fetching it.
        """
        if self.assignments:
            if only_approved:
                for a in self.assignments:
                    if a.manager_accepts:
                        return a.supplier
            else:
                return [a.supplier for a in self.assignments]
        return None

    def is_production_in_house(self):
        """ Return True if one of the job assignment suppliers is a manager """
        return any(a.supplier.has_authority(self.domain.id, [ADMIN, MANAGER])
                   for a in self.assignments)

    def in_house_assignment(self):
        """ Return an assignment if the assignment is in-house """
        if self.is_production_in_house():
            return self.assignments[0]
        return None

    def accepted_assignment(self):
        """ Return the production assignment accepted by the manager """
        for assignment in self.assignments:
            if assignment.manager_accepts:
                return assignment
        return None

    def assignment_deadline(self):
        """ Return the assignment deadline from one of the assignments """
        if len(self.assignments) > 0:
            return self.assignments[0].assignment_deadline
        return None

    def assignment_by_supplier_id(self, user_id):
        """ Return a job assignment associated with
            the user provided and return None if no assignment can
            be found with the user.
        """
        for assignment in self.assignments:
            if assignment.supplier.id == user_id:
                return assignment
        return None

    def production_shipment(self):
        """ Return the shipment object which has been created for production
            procedure.
        """
        for assignment in self.assignments:
            if assignment.shipment:
                return assignment.shipment
        return None

    def production_display_info(self):
        """ Return the PRODUCTION_STATUS dict from
            sdk.python.util.status with the values based on the
            job.production_status
        """
        return status_info(self.production_status,
                           PRODUCTION_STATUS, "production ")

    def design_display_info(self):
        """ Return the DESIGN_STATUS dict from
            sdk.python.util.status with the values based on the
            job.design_status
        """
        return status_info(self.design_status, DESIGN_STATUS, "design ")

    def payment_display_info(self):
        """ Return the PAYMENT_STATUS dict from
            sdk.python.util.status with the values based on the
            job.payment_status
        """
        return status_info(self.payment_status, PAYMENT_STATUS, "payment ")

    def shipment_display_info(self):
        """ Return the SHIPPING_STATUS dict from sdk.python.util.status with the values
            based on the job.shipping_status
        """
        return status_info(self.shipping_status, SHIPPING_STATUS, "shipment ")

    def supplier_shipment_info(self):
        """ Return the shipping information submitted buy the supplier """
        assignment = self.accepted_assignment()
        if assignment:
            return assignment.shipment
        return None

    def suppliers_unarchived_assignment(self, user_id):
        """ Returns an unarchived assignment of the given user id if such
            assignment exists
        """
        for assignment in self.assignments:
            if assignment.supplier.id == user_id and not assignment.archived:
                return assignment
        return None

    def product_unit_price(self):
        """ Display the product unit_price. This value takes into consideration
            discounted unit price based on server calculated product unit price
            after it got discounted
        """
        return self.cost_per_unit

    def assignment_earliest_production(self):
        """ Return the assignment which is able to get the
            production completed in the fastest amount of time
        """
        earliest_deadline = None
        earliest_supplier = None
        for assignment in self.assignments:
            if assignment.quote:
                agreed_deadline = assignment.quote.agreed_deadline
                if not earliest_deadline or agreed_deadline < earliest_deadline:
                    earliest_deadline = assignment.quote.agreed_deadline
                    earliest_supplier = assignment
        return earliest_supplier

    def assignment_lowest_quote(self):
        """ Return the lowest quote out of all the assignments """
        lowest_quote = None
        lowest_supplier = None
        for assignment in self.assignments:
            if assignment.quote:
                if not lowest_quote or \
                        assignment.quote.quote_total() < lowest_quote:
                    lowest_quote = assignment.quote.quote_total()
                    lowest_supplier = assignment
        return lowest_supplier

    def get_section_name(self, section):
        """ Generate section name that can be used to generate private attribute
            to store notifications info of different sections in job.
        """
        fmt = "_{}_notifications_"
        if not section:
            return fmt.format("all")
        return fmt.format(JOB_SECTION_STRINGS[section])

    def user_job_notifications(self, user):
        """ Return job notifications which the current user is the recipient """
        notes = self.notifications
        return [n for n in notes if n.recipient and n.recipient.id == user.id]

    def user_unseen_job_notifications(self, user):
        """ Return all the job notifications of the provided user as the
            recipient if the notification has not been marked as seen
        """
        return [n for n in self.user_job_notifications(user) if not n.seen]

    def user_unseen_job_notifications_count(self, user):
        """ Return the length of the user_unseen_job_notifications method """
        return len(self.user_unseen_job_notifications(user))

    def unseen_notifications_by_section(self, user, section_int):
        """ Return all job related notifications of the current user
            as the recipient which have not been marked as 'seen' and which
            are in provided section
        """
        notes = self.user_unseen_job_notifications(user)
        return [n for n in notes if section_int == n.section]

    def unseen_notifications_by_section_count(self, user, section_int):
        """ Return the length of the unseen_notifications_by_section method """
        return len(self.unseen_notifications_by_section(user, section_int))

    def shipping_finished(self):
        """ Return whether the shipping has been finished """
        return self.shipping_status >= \
            SHIPPING_STATUS["DISPATCHED"]["dbValue"] or \
            not self.needs_shipping

    def drafting_finished(self):
        """ Return whether the drafting has been finished """
        return self.design_status >= \
            DESIGN_STATUS["DRAFTING_APPROVED"]["dbValue"] or \
            not self.needs_drafting

    def production_finished(self):
        """ Return whether the production has been finished """
        return (self.production_status is not None and
                self.production_status >=
                PRODUCTION_STATUS["SHIPPED"]["dbValue"]) or \
            not self.needs_production

    def payment_finished(self):
        """ Return whether the payment has been finished """
        if self.payment_status:
            return self.payment_status >= \
                PAYMENT_STATUS["FULLY_PAID"]["dbValue"]

    def is_ready_to_complete(self):
        """ Return whether the job is ready for move to the
            completed stage
        """
        return self.drafting_finished() and self.production_finished() and\
            self.payment_finished() and self.shipping_finished()

    def dictionary_of_related_address_names_and_ids(self, job_shipping=True,
                                                    production_shipping=True,
                                                    manager_addresses=True,
                                                    client_addresses=True):
        """ Return a list of the related job address ids and names.  """
        saved_addresses = []
        if self.shipping and job_shipping:
            saved_addresses.append({'name': 'Job Shipping Address',
                                    'id': self.shipping.id})

        if self.production_shipping_address and production_shipping:
            saved_addresses.append({'name': 'Job Shipping Address',
                                    'id': self.production_shipping_address.id})

        if self.client and client_addresses:
            saved_addresses += self.client. \
                dictionary_of_addresses_and_ids("Client Address")
        if self.manager and manager_addresses:
            saved_addresses += self.manager. \
                dictionary_of_addresses_and_ids("Manager Address")

        return saved_addresses

    def draft_time_line(self, reverse=True):
        """ Return job draft related object in chronological order
            as a dictionary. This is used for creating a job draft timeline.
        """
        time_line_dict = {}

        for comment in self.draft_comments:
            time_line_dict[comment.date] = \
                events.Event(events.DRAFT_COMMENT, comment)
        for draft in self.drafts:
            time_line_dict[draft.date] = \
                events.Event(events.DRAFT_UPLOADED, draft)
            if draft.accepted:
                time_line_dict[draft.accepted] = \
                    events.Event(events.DRAFT_APPROVED, draft.accepted)
            if draft.resend_date:
                time_line_dict[draft.resend_date] = \
                    events.Event(events.DRAFT_RESENT, draft.resend_date)
            for comment in draft.comments:
                if comment.change_request:
                    time_line_dict[comment.date] = \
                        events.Event(events.DRAFT_REJECTED, comment)
                else:
                    time_line_dict[comment.date] = \
                        events.Event(events.DRAFT_COMMENT, comment)

        return collections.OrderedDict(sorted(time_line_dict.items(),
                                              reverse=reverse,
                                              key=lambda t: t[0]))

    def job_info_time_line(self, reverse=True):
        """ Return the job comments in reverse with the youngest at
            bottom of the list.
        """
        time_line_dict = {}
        for comment in self.comments:
            time_line_dict[comment.date] = \
                events.Event(events.JOB_COMMENT, comment)
        return collections.OrderedDict(sorted(time_line_dict.items(),
                                              reverse=reverse,
                                              key=lambda t: t[0]))

    def is_complete_string(self):
        """ Return a string 'Completed' if the job is complete
            else it will return a string 'Uncompleted'.
        """
        if self.completed:
            return "Completed"
        return "Uncompleted"

    def can_be_archived_by(self, user):
        """ Check if the job can be archived by the user making the request. """
        return self.rights.delete

    def total_inc_tax(self):
        """ Return the job cost + the tax amount """
        return self.tax_amount + self.cost


class Jobs(sdk.python.entities.Resource):

    entity_class = Job
    json_name = 'jobs'


jobs = Jobs()
